dartagnan/src/main/java/com/dat3m/dartagnan/expression/type/AggregateType.java

```diff
@@ -2,33 +2,61 @@
 
 import com.dat3m.dartagnan.expression.Type;
 
+import java.util.ArrayList;
 import java.util.List;
 import java.util.stream.Collectors;
+import java.util.stream.IntStream;
 
-public final class AggregateType implements Type {
+import static com.dat3m.dartagnan.expression.type.TypeFactory.paddedSize;
 
-    private final List<Type> fields;
+public class AggregateType implements Type {
 
-    AggregateType(List<Type> directFields) {
-        this.fields = List.copyOf(directFields);
+    private static final TypeFactory types = TypeFactory.getInstance();
+
+    private final List<TypeOffset> directFields;
+
+    AggregateType(List<Type> fields) {
+        this(fields, computeDefaultOffsets(fields));
+    }
+
+    AggregateType(List<Type> fields, List<Integer> offsets) {
+        this.directFields = IntStream.range(0, fields.size()).boxed().map(i -> new TypeOffset(fields.get(i), offsets.get(i))).toList();
+    }
+
+    private static List<Integer> computeDefaultOffsets(List<Type> fields) {
+        List<Integer> offsets = new ArrayList<>();
+        int maxAlignment = 0;
+        int maxOffset = 0;
+        if (!fields.isEmpty()) {
+            maxAlignment = types.getAlignment(fields.get(0));
+            maxOffset = types.getMemorySizeInBytes(fields.get(0));
+            offsets.add(0);
+        }
+        for (int i = 1; i < fields.size(); i++) {
+            maxAlignment = Integer.max(types.getAlignment(fields.get(i)), maxAlignment);
+            maxOffset = paddedSize(maxOffset, maxAlignment);
+            offsets.add(maxOffset);
+            maxOffset += types.getMemorySizeInBytes(fields.get(i));
+        }
```

Comment:
This computation looks wrong to me. E.g. `struct { int32, int8, int8, int8 }` should have a size of 8 bytes with offsets 0, 4, 5, and 6. It seems to me that your computation will force 4-byte alignment on the `int8` members because of the 4-byte member in the beginning.